csharp/src/Ice/TcpAcceptor.cs

```diff
@@ -16,12 +19,33 @@ internal class TcpAcceptor : IAcceptor
         private readonly IConnectionManager _manager;
         private readonly Socket _socket;
         private readonly IPEndPoint _addr;
+        private const byte TlsHandshakeRecord = 0x16;
 
         public async ValueTask<Connection> AcceptAsync()
         {
             Socket fd = await _socket.AcceptAsync().ConfigureAwait(false);
 
-            ITransceiver transceiver = ((TcpEndpoint)Endpoint).CreateTransceiver(fd, _adapter.Name);
+            bool secure = Endpoint.IsAlwaysSecure || !_adapter.AcceptNonSecure;
+
+            if (_adapter.Protocol == Protocol.Ice2 && _adapter.AcceptNonSecure)
+            {
+                Debug.Assert(_adapter.Communicator.ConnectTimeout != TimeSpan.Zero);
+                // TODO: we are using reusing ConnectTimeout here so that peeking cannot block forever.
+                // However, this means that it's possible to end up waiting 2 * ConnectTime if reading the
+                // first byte is slow and then the actual connection initialization is also slow.
+
+                using var source = new CancellationTokenSource(_adapter.Communicator.ConnectTimeout);
+
+                // Peek one byte into the tcp stream to see if it contains the TLS handshake record
+                var buffer = new ArraySegment<byte>(new byte[1]);
```

Comment:
> if it's a secure connection, we discard this buffered transceiver - or maybe we can somehow reuse it under the SslTransceiver?

It's possible to pass something else than a C# NetworkStream to the C# SslStream object... so in theory it would be possible but it sounds like going through lots of trouble for very little gain.